misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,461 @@
+    # plt.title(title, fontsize='x-small' if small else 'medium')  # medium is the default
+    plt.title(title)
+    plt.grid(True)
+    # X Ticks: center of bars rather than to the left
+    # todo: why ticks & labels getting super bolded with this fix?
+    if not is_timeseries and xtick_fix:
+        plt.xticks([edge + 0.5 for edge in data[:-1]], data[:-1])
+    if xtick_step:
+        plt.xticks(range(0, int(max(data)), xtick_step), rotation=45)
+
+    if small:
+        # Font size
+        # todo: this isn't working; not a big priority
+        plt.rcParams["font.size"] = "6" if len(title) < 30 else "6"  # long titles make plot smaller
+        # Axis
+        if not is_timeseries:  # it makes sense to get rid of axis for timeseries, but not otherwise
+            plt.tick_params(axis='y', which='both', left=False, labelleft=False)  # Turn off y-axis ticks and labels
+        else:
+            plt.axis('off')
+        # Save
+        # plt.savefig(OUTDIR / f'{filename}.png', dpi=100)
+        plt.savefig(OUTDIR / f'{filename} - small.png', bbox_inches='tight', dpi=100)
+    else:
+        # Labels
+        plt.xlabel(xlab)
+        plt.ylabel(ylab)
+        # Save
+        plt.savefig(OUTDIR / f'{filename}.png', dpi=100)
+    plt.clf()
+
+
+# todo: histogram sections duplicative. could be refactored to be more DRY
+def plots(df: pd.DataFrame, df_dev0: pd.DataFrame, small=False, dev_data_plots=False):
+    """Plots"""
+    # todo: consider dev + non-dev in same plot
+    # Histogram: Value set n selections
+    # - Size distribution histogram. How many code sets in initial call of log session
+    datasets = ((df, 'Dev data included'), (df_dev0, '')) if dev_data_plots else ((df_dev0, ''),)
+    for df_i, name_suffix in datasets:
+        # Title
+        title = f'Code set comparison size{f" - {name_suffix}" if name_suffix else ""}'
+        # Select data
+        df_i2 = df_i[df_i['api_call'] == 'get-csets']
+        df_i2['len_codeset_ids'] = df_i['codeset_ids'].apply(lambda x: len(x))
+        data: List[int] = list(df_i2['len_codeset_ids'])
+        bins = range(min(data), max(data) + 2)
+        # Render
+        plot_render(data, title, 'Number of code sets being compared', 'Frequency', bins, small)
+
+    # Histogram: API calls per week
+    for df_i, name_suffix in datasets:
+        # Title
+        title = f'API calls per week{f" - {name_suffix}" if name_suffix else ""}'
+        # Select data
+        # for df, name_suffix in ((df_apiruns, 'Dev data included'), (df_apiruns_dev0, '')):
+        data: List = list(df_i['week'])
+        bins = len(set(data))
+        # Render
+        plot_render(data, title, 'Week', 'API calls', bins, small, is_timeseries=True)
+
+    # Histogram: Session duration (duration_seconds)
+    for df_i, name_suffix in datasets:
+        # Title
+        title = f'Duration, session API calls{f" - {name_suffix}" if name_suffix else ""}'
```

Comment:
## Duration, session API calls
There's a whole lot of variation in the <0 range, but this is good enough I think. Adding more bins could make us see that better though.

![duration_session_api_calls_log_scale - small](https://github.com/jhu-bids/TermHub/assets/13045020/06930a5b-f728-4f6b-821e-03ff98c6c230)

<details><summary>Big version</summary>
<p>

![duration_session_api_calls_log_scale](https://github.com/jhu-bids/TermHub/assets/13045020/0c211ccb-5fc8-44f7-bb52-2865fc6a78e4)

</p>
</details> 

### `duration_seconds` vs `process_seconds`
@Sigfried I was a little unsure of what to use here. I think you intended for me to use `duration_seconds`. It was consistent with what I was able to calculate from the timestamps in each API call group. However, some things I wonder:

1. Was it indeed better that I chose `duration_seconds`?

2. What is `process_seconds`, and why is it different than `duration_seconds`?
Does it have something to do with async? Like are multiple things being processed simultaneously so that the sum time of each call doesn't equate with how long it takes for the user to get all the data?

3. Is there an issue with `duration_seconds` calculation from timestamps?
I'm concerned there might be a problem with this calculation. I think it would be better for each API call to actually have 2 timestamps: start and end time. Then, we could consider that the duration of an API call group would be the (end time of last call) - (start time of the 1st call). However, given that each call only has 1 timestamp, we can't calculate this, and I am assuming each timestamp represents the start time (or end time) only. If that's the case, then I think the duration of the final call is not being included in `duration_seconds`.